File: examples/dekita/.claude/hooks/systematization_check.py
#!/usr/bin/env python3
"""セッション終了時に教訓が仕組み化されたか確認する。

Why:
    教訓をドキュメント化しただけでは、同じ問題が再発する。
    フック/CI/ツールによる仕組み化を強制することで、再発を防止する。

What:
    - 教訓パターン検出（教訓、反省点、学び、気づき、lesson learned等）
    - 仕組み化ファイル変更検出（.claude/hooks/*.py、.github/workflows/*.yml等）
    - 教訓あり＆仕組み化なしの場合にブロック
    - 誤検知緩和（複数インジケータ要求、明示的スキップ許可）

Remarks:
    - problem-report-checkはIssue作成を確認、本フックは仕組み化を確認
    - 強パターン（「仕組み化が必要」等）は単独でもトリガー
    - 「仕組み化しました」等の完了パターンは誤検知として除外

Changelog:
    - silenvx/dekita#468: フック追加
"""

import json
import re
import sys

from lib.execution import log_hook_execution
from lib.session import parse_hook_input
from lib.transcript import load_transcript

# Lesson/learning patterns in Japanese
LESSON_PATTERNS_JA = [
    r"教訓として",
    r"反省点として",
    r"反省点は",
    r"学びとして",
    r"学んだこと",
    r"気づいた(?:こと|点)",
    r"今後は.*(?:する|しない)(?:べき|必要)",
    r"(?:再発)?防止(?:策)?(?:として|は)",
    r"次回(?:から)?は",
    r"改善(?:点|策|が必要)",
]

# Lesson/learning patterns in English
LESSON_PATTERNS_EN = [
    r"lesson(?:s)?\s+learned",
    r"key\s+takeaway",
    r"should\s+have\s+(?!been\s+(?:done|completed))",  # "should have" but not completion
    r"in\s+the\s+future",
    r"next\s+time",
    r"going\s+forward",
    r"to\s+prevent\s+this",
    r"to\s+avoid\s+this",
]

ALL_LESSON_PATTERNS = LESSON_PATTERNS_JA + LESSON_PATTERNS_EN

# Strong indicators that definitely need systematization
STRONG_LESSON_PATTERNS = [
    r"仕組み化(?:する|が必要|すべき)",
    r"hook(?:を|で|が)",
    r"フック(?:を|で|が)",
    r"CI(?:で|に|を)",
    r"自動化(?:する|が必要|すべき)",
]

# Patterns that indicate false positives
FALSE_POSITIVE_PATTERNS = [
    r"仕組み化(?:しました|済み|完了)",
    r"hook(?:を)?(?:作成|追加|実装)(?:しました|済み|完了)",
    r"フック(?:を)?(?:作成|追加|実装)(?:しました|済み|完了)",
    r"対応不要",
    r"アクション不要",
    r"no\s+action\s+needed",
    r"already\s+(?:implemented|done|addressed)",
]

# Systematization file patterns
SYSTEMATIZATION_PATTERNS = [
    r"\.claude/hooks/.*\.py$",
    r"\.github/workflows/.*\.ya?ml$",
    r"\.claude/scripts/.*\.(?:py|sh)$",
]


def extract_claude_messages(transcript: list[dict]) -> list[str]:
    """Extract text from Claude's messages in the transcript."""
    messages = []
    for entry in transcript:
        if entry.get("role") == "assistant":
            content = entry.get("content", [])
            if isinstance(content, list):
                for block in content:
                    if isinstance(block, dict) and block.get("type") == "text":
                        text = block.get("text", "")
                        if text:
                            messages.append(text)
            elif isinstance(content, str):
                messages.append(content)
    return messages


def extract_file_operations(transcript: list[dict]) -> list[str]:
    """Extract file paths from Edit/Write tool uses."""
    files = []
    for entry in transcript:
        if entry.get("role") == "assistant":
            content = entry.get("content", [])
            if isinstance(content, list):
                for block in content:
                    if isinstance(block, dict) and block.get("type") == "tool_use":
                        tool_name = block.get("name", "")
                        if tool_name in ("Edit", "Write"):
                            file_path = block.get("input", {}).get("file_path", "")
                            if file_path:
                                files.append(file_path)
    return files


def has_false_positive(messages: list[str]) -> bool:
    """Check if any message indicates a false positive."""
    for msg in messages:
        for pattern in FALSE_POSITIVE_PATTERNS:
            if re.search(pattern, msg, re.IGNORECASE):
                return True
    return False


def find_lesson_patterns(messages: list[str]) -> tuple[list[str], bool]:
    """Find lesson patterns in Claude's messages.

    Returns:
        Tuple of (list of lesson excerpts, has_strong_pattern)
    """
    lessons = []
    has_strong = False

    for msg in messages:
        # Check strong patterns first
        for pattern in STRONG_LESSON_PATTERNS:
            if re.search(pattern, msg, re.IGNORECASE):
                has_strong = True

        # Check normal lesson patterns
        for pattern in ALL_LESSON_PATTERNS:
            match = re.search(pattern, msg, re.IGNORECASE)
            if match:
                # Extract surrounding context
                start = max(0, match.start() - 30)
                end = min(len(msg), match.end() + 30)
                excerpt = msg[start:end].strip()
                if start > 0:
                    excerpt = "..." + excerpt
                if end < len(msg):
                    excerpt = excerpt + "..."
                lessons.append(excerpt)
                break  # Only count once per message

    return lessons, has_strong


def find_systematization_files(files: list[str]) -> list[str]:
    """Find files that indicate systematization."""
    systematized = []
    for file_path in files:
        for pattern in SYSTEMATIZATION_PATTERNS:
            if re.search(pattern, file_path):
                systematized.append(file_path)
                break
    return systematized


def main():
    """Stop hook to detect lessons learned and request systematization.

    Uses exit code 0 with ACTION_REQUIRED message to allow continuation.
    Claude Code will see the message and take action autonomously.
    """
    result = {"decision": "approve"}

    try:
        input_data = parse_hook_input()

        # Skip if stop hook is already active (prevents infinite loops)
        # When stop_hook_active=True, exit 0 to allow stop and prevent infinite retry
        if input_data.get("stop_hook_active"):
            log_hook_execution("systematization-check", "approve", "stop_hook_active")
            print(json.dumps(result))
            return

        transcript_path = input_data.get("transcript_path")
        if not transcript_path:
            log_hook_execution("systematization-check", "approve", "no transcript path")
            print(json.dumps(result))
            return

        transcript = load_transcript(transcript_path)
        if not transcript:
            log_hook_execution("systematization-check", "approve", "transcript load failed")
            print(json.dumps(result))
            return

        claude_messages = extract_claude_messages(transcript)
        file_operations = extract_file_operations(transcript)

        # Check for false positives first
        if has_false_positive(claude_messages):
            log_hook_execution("systematization-check", "approve", "false positive detected")
            print(json.dumps(result))
            return

        # Find lessons and systematization
        lessons, has_strong = find_lesson_patterns(claude_messages)
        systematized_files = find_systematization_files(file_operations)

        # Decision logic:
        # - If strong lesson pattern AND no systematization -> force continue with exit 2
        # - If 2+ normal lesson patterns AND no systematization -> force continue with exit 2
        # - Otherwise -> approve with warning if lessons detected

        should_block = False
        if has_strong and not systematized_files:
            should_block = True
        elif len(lessons) >= 2 and not systematized_files:
            should_block = True

        if should_block:
            excerpts = lessons[:3]
            excerpt_text = "\n".join(f"  - {e[:80]}" for e in excerpts)
            if len(lessons) > 3:
                excerpt_text += f"\n  - ... 他 {len(lessons) - 3} 件"

            # ACTION_REQUIRED format for Claude Code to take autonomous action
            reason = (
                f"[ACTION_REQUIRED: SYSTEMATIZATION]\n"
                f"教訓が見つかりましたが、仕組み化されていません。\n\n"
                f"**検出された教訓** ({len(lessons)}件):\n{excerpt_text}\n\n"
                "Claude Codeは以下のいずれかを実行してください:\n"
                "1. `.claude/hooks/` にフックを作成してブロック機構を実装\n"
                "2. `.github/workflows/` にCIチェックを追加\n"
                "3. `.claude/scripts/` にスクリプトを作成\n"
                "4. 上記が不要な場合はIssueを作成して理由を記録\n\n"
                "ドキュメント（AGENTS.md等）への追記だけでは不十分です。"
            )
            log_hook_execution(
                "systematization-check",
                "warn",
                f"lessons: {len(lessons)}, strong: {has_strong}, files: {len(systematized_files)}",
            )
            # Issue #2026: Use exit code 0 to allow continuation
            # Print ACTION_REQUIRED to stderr for Claude Code to see and act on
            print(f"[systematization-check] {reason}", file=sys.stderr)
            # Continue with approve instead of blocking
        elif lessons:
            # Lessons detected but systematization also detected
            result["systemMessage"] = (
                f"✅ [systematization-check] 教訓検出: {len(lessons)}件, "
                f"仕組み化ファイル: {len(systematized_files)}件"
            )
            log_hook_execution(
                "systematization-check",
                "approve",
                f"lessons: {len(lessons)}, files: {len(systematized_files)}",
            )
        else:
            log_hook_execution("systematization-check", "approve", "no lessons detected")

    except Exception as e:
        log_hook_execution("systematization-check", "approve", f"error: {e}")
        print(f"[systematization-check] Error: {e}", file=sys.stderr)

    print(json.dumps(result))


if __name__ == "__main__":
    main()

File: examples/dekita/.claude/hooks/handlers/systematization_check.ts
#!/usr/bin/env bun
/**
 * セッション終了時に教訓が仕組み化されたか確認する。
 *
 * Why:
 *   教訓をドキュメント化しただけでは、同じ問題が再発する。
 *   フック/CI/ツールによる仕組み化を強制することで、再発を防止する。
 *
 * What:
 *   - 教訓パターン検出（教訓、反省点、学び、気づき、lesson learned等）
 *   - 仕組み化ファイル変更検出（.claude/hooks/*.py、.github/workflows/*.yml等）
 *   - 教訓あり＆仕組み化なしの場合にブロック
 *   - 誤検知緩和（複数インジケータ要求、明示的スキップ許可）
 *
 * Remarks:
 *   - problem-report-checkはIssue作成を確認、本フックは仕組み化を確認
 *   - 強パターン（「仕組み化が必要」等）は単独でもトリガー
 *   - 「仕組み化しました」等の完了パターンは誤検知として除外
 *   - Python版: systematization_check.py
 *
 * Changelog:
 *   - silenvx/dekita#468: フック追加
 *   - silenvx/dekita#2986: TypeScript版に移植
 */

import { formatError } from "../lib/format_error";
import { logHookExecution } from "../lib/logging";
import { parseHookInput } from "../lib/session";
import { loadTranscript } from "../lib/transcript";

const HOOK_NAME = "systematization-check";

// Lesson/learning patterns in Japanese
const LESSON_PATTERNS_JA = [
  /教訓として/i,
  /反省点として/i,
  /反省点は/i,
  /学びとして/i,
  /学んだこと/i,
  /気づいた(?:こと|点)/i,
  /今後は.*(?:する|しない)(?:べき|必要)/i,
  /(?:再発)?防止(?:策)?(?:として|は)/i,
  /次回(?:から)?は/i,
  /改善(?:点|策|が必要)/i,
];

// Lesson/learning patterns in English
const LESSON_PATTERNS_EN = [
  /lesson(?:s)?\s+learned/i,
  /key\s+takeaway/i,
  /should\s+have\s+(?!been\s+(?:done|completed))/i, // "should have" but not completion
  /in\s+the\s+future/i,
  /next\s+time/i,
  /going\s+forward/i,
  /to\s+prevent\s+this/i,
  /to\s+avoid\s+this/i,
];

const ALL_LESSON_PATTERNS = [...LESSON_PATTERNS_JA, ...LESSON_PATTERNS_EN];

// Strong indicators that definitely need systematization
const STRONG_LESSON_PATTERNS = [
  /仕組み化(?:する|が必要|すべき)/i,
  /hook(?:を|で|が)/i,
  /フック(?:を|で|が)/i,
  /CI(?:で|に|を)/i,
  /自動化(?:する|が必要|すべき)/i,
];

// Patterns that indicate false positives
const FALSE_POSITIVE_PATTERNS = [
  /仕組み化(?:しました|済み|完了)/i,
  /hook(?:を)?(?:作成|追加|実装)(?:しました|済み|完了)/i,
  /フック(?:を)?(?:作成|追加|実装)(?:しました|済み|完了)/i,
  /対応不要/i,
  /アクション不要/i,
  /no\s+action\s+needed/i,
  /already\s+(?:implemented|done|addressed)/i,
];

// Systematization file patterns
const SYSTEMATIZATION_FILE_PATTERNS = [
  /\.claude\/hooks\/.*\.py$/,
  /\.github\/workflows\/.*\.ya?ml$/,
  /\.claude\/scripts\/.*\.(?:py|sh)$/,
  /\.claude\/hooks\/handlers\/.*\.ts$/,
];

interface ContentBlock {
  type?: string;
  text?: string;
  name?: string;
  input?: Record<string, unknown>;
  [key: string]: unknown;
}

interface TranscriptEntry {
  role?: string;
  content?: string | ContentBlock[];
  [key: string]: unknown;
}

/**
 * Extract text from Claude's messages in the transcript.
 */
export function extractClaudeMessages(transcript: TranscriptEntry[]): string[] {
  const messages: string[] = [];

  for (const entry of transcript) {
    if (entry.role === "assistant") {
      const content = entry.content;
      if (Array.isArray(content)) {
        for (const block of content) {
          if (block && typeof block === "object" && block.type === "text" && block.text) {
            messages.push(block.text);
          }
        }
      } else if (typeof content === "string") {
        messages.push(content);
      }
    }
  }

  return messages;
}

/**
 * Extract file paths from Edit/Write tool uses.
 */
export function extractFileOperations(transcript: TranscriptEntry[]): string[] {
  const files: string[] = [];

  for (const entry of transcript) {
    if (entry.role === "assistant") {
      const content = entry.content;
      if (Array.isArray(content)) {
        for (const block of content) {
          if (block && typeof block === "object" && block.type === "tool_use") {
            const toolName = block.name ?? "";
            if (toolName === "Edit" || toolName === "Write") {
              const input = block.input as Record<string, unknown> | undefined;
              const filePath = input?.file_path;
              if (typeof filePath === "string") {
                files.push(filePath);
              }
            }
          }
        }
      }
    }
  }

  return files;
}

/**
 * Check if any message indicates a false positive.
 */
export function hasFalsePositive(messages: string[]): boolean {
  for (const msg of messages) {
    for (const pattern of FALSE_POSITIVE_PATTERNS) {
      if (pattern.test(msg)) {
        return true;
      }
    }
  }
  return false;
}

/**
 * Find lesson patterns in Claude's messages.
 */
export function findLessonPatterns(messages: string[]): { lessons: string[]; hasStrong: boolean } {
  const lessons: string[] = [];
  let hasStrong = false;

  for (const msg of messages) {
    // Check strong patterns first
    for (const pattern of STRONG_LESSON_PATTERNS) {
      if (pattern.test(msg)) {
        hasStrong = true;
        break;
      }
    }

    // Check normal lesson patterns
    for (const pattern of ALL_LESSON_PATTERNS) {
      const match = msg.match(pattern);
      if (match?.index !== undefined) {
        // Extract surrounding context
        const start = Math.max(0, match.index - 30);
        const end = Math.min(msg.length, match.index + match[0].length + 30);
        let excerpt = msg.slice(start, end).trim();
        if (start > 0) {
          excerpt = `...${excerpt}`;
        }
        if (end < msg.length) {
          excerpt = `${excerpt}...`;
        }
        lessons.push(excerpt);
        break; // Only count once per message
      }
    }
  }

  return { lessons, hasStrong };
}

/**
 * Find files that indicate systematization.
 */
export function findSystematizationFiles(files: string[]): string[] {
  const systematized: string[] = [];

  for (const filePath of files) {
    for (const pattern of SYSTEMATIZATION_FILE_PATTERNS) {
      if (pattern.test(filePath)) {
        systematized.push(filePath);
        break;
      }
    }
  }

  return systematized;
}

async function main(): Promise<void> {
  const result: { decision?: string; systemMessage?: string } = {};

  let sessionId: string | undefined;

  try {
    const inputData = await parseHookInput();
    sessionId = inputData.session_id;

    // Skip if stop hook is already active (prevents infinite loops)
    if (inputData.stop_hook_active) {
      await logHookExecution(HOOK_NAME, "approve", "stop_hook_active", undefined, { sessionId });
      console.log(JSON.stringify(result));
      return;
    }

    const transcriptPath = inputData.transcript_path as string | undefined;
    if (!transcriptPath) {
      await logHookExecution(HOOK_NAME, "approve", "no transcript path", undefined, { sessionId });
      console.log(JSON.stringify(result));
      return;
    }

    const transcript = loadTranscript(transcriptPath);
    if (!transcript) {
      await logHookExecution(HOOK_NAME, "approve", "transcript load failed", undefined, {
        sessionId,
      });
      console.log(JSON.stringify(result));
      return;
    }

    const claudeMessages = extractClaudeMessages(transcript);
    const fileOperations = extractFileOperations(transcript);

    // Check for false positives first
    if (hasFalsePositive(claudeMessages)) {
      await logHookExecution(HOOK_NAME, "approve", "false positive detected", undefined, {
        sessionId,
      });
      console.log(JSON.stringify(result));
      return;
    }

    // Find lessons and systematization
    const { lessons, hasStrong } = findLessonPatterns(claudeMessages);
    const systematizedFiles = findSystematizationFiles(fileOperations);

    // Decision logic:
    // - If strong lesson pattern AND no systematization -> warn with ACTION_REQUIRED
    // - If 2+ normal lesson patterns AND no systematization -> warn with ACTION_REQUIRED
    // - Otherwise -> approve with message if lessons detected

    let shouldWarn = false;
    if (hasStrong && systematizedFiles.length === 0) {
      shouldWarn = true;
    } else if (lessons.length >= 2 && systematizedFiles.length === 0) {
      shouldWarn = true;
    }

    if (shouldWarn) {
      const excerpts = lessons.slice(0, 3);
      let excerptText = excerpts.map((e) => `  - ${e.slice(0, 80)}`).join("\n");
      if (lessons.length > 3) {
        excerptText += `\n  - ... 他 ${lessons.length - 3} 件`;
      }

      // ACTION_REQUIRED format for Claude Code to take autonomous action
      const reason = `[ACTION_REQUIRED: SYSTEMATIZATION]\n教訓が見つかりましたが、仕組み化されていません。\n\n**検出された教訓** (${lessons.length}件):\n${excerptText}\n\nClaude Codeは以下のいずれかを実行してください:\n1. \`.claude/hooks/\` にフックを作成してブロック機構を実装\n2. \`.github/workflows/\` にCIチェックを追加\n3. \`.claude/scripts/\` にスクリプトを作成\n4. 上記が不要な場合はIssueを作成して理由を記録\n\nドキュメント（AGENTS.md等）への追記だけでは不十分です。`;

      await logHookExecution(
        HOOK_NAME,
        "warn",
        `lessons: ${lessons.length}, strong: ${hasStrong}, files: ${systematizedFiles.length}`,
        undefined,
        { sessionId },
      );

      // Print ACTION_REQUIRED to stderr for Claude Code to see and act on
      console.error(`[${HOOK_NAME}] ${reason}`);
      // Continue with approve instead of blocking (Issue #2026)
    } else if (lessons.length > 0) {
      // Lessons detected but systematization also detected
      result.systemMessage = `✅ [${HOOK_NAME}] 教訓検出: ${lessons.length}件, 仕組み化ファイル: ${systematizedFiles.length}件`;
      await logHookExecution(
        HOOK_NAME,
        "approve",
        `lessons: ${lessons.length}, files: ${systematizedFiles.length}`,
        undefined,
        { sessionId },
      );
    } else {
      await logHookExecution(HOOK_NAME, "approve", "no lessons detected", undefined, { sessionId });
    }
  } catch (error) {
    await logHookExecution(HOOK_NAME, "approve", `error: ${formatError(error)}`, undefined, {
      sessionId,
    });
    console.error(`[${HOOK_NAME}] Error: ${formatError(error)}`);
  }

  console.log(JSON.stringify(result));
}

if (import.meta.main) {
  main();
}
